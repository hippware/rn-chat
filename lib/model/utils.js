"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const mobx_1 = require("mobx");
function waitFor(condition) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => {
            mobx_1.when(() => {
                let res = false;
                try {
                    res = condition();
                }
                catch (e) {
                    reject(e);
                }
                return res;
            }, () => {
                resolve();
            });
        });
    });
}
exports.waitFor = waitFor;
//# sourceMappingURL=utils.js.map