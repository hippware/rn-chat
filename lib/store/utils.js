"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const mobx_1 = require("mobx");
function pad(n, width, z = '0') {
    const str = n.toString();
    return str.length >= width ? str : new Array(width - str.length + 1).join(z) + n;
}
function process(result) {
    if (typeof result === 'object') {
        if (Array.isArray(result)) {
            return result.map(el => process(el));
        }
        else {
            if (result['#text'] && Object.keys(result).length === 1) {
                return result['#text'];
            }
            const res = {};
            let changed = false;
            // eslint-disable-next-line
            for (const key in result) {
                // eslint-disable-next-line
                if (result.hasOwnProperty(key)) {
                    changed = true;
                    res[key] = process(result[key]);
                }
            }
            if (!changed) {
                return undefined;
            }
            return res;
        }
    }
    else {
        return result;
    }
}
function isArray(res) {
    return Array.isArray(res) || mobx_1.isObservableArray(res);
}
exports.isArray = isArray;
function camelize(str) {
    return str
        .replace(/\W|_|\d/g, ' ')
        .replace(/(?:^\w|[A-Z]|\b\w)/g, (letter, index) => {
        return index === 0 ? letter.toLowerCase() : letter.toUpperCase();
    })
        .replace(/\s+/g, '');
}
exports.camelize = camelize;
function processMap(data) {
    const res = {};
    Object.keys(data).forEach(key => {
        const value = data[key];
        try {
            if (value && value !== 'null' && key !== 'field') {
                if (key === 'roles') {
                    res.roles = isArray(data.roles.role) ? data.roles.role : [data.roles.role];
                }
                else if (['followers', 'bots', 'followed'].indexOf(key) !== -1) {
                    res[key + 'Size'] = parseInt(data[key].size);
                }
                else if (data[key].thumbnail_url !== undefined) {
                    // we have image here!
                    if (data[key]['#text']) {
                        res[key] = { id: data[key]['#text'], url: data[key].thumbnail_url };
                    }
                }
                else if (key === 'subscribed') {
                    res.isSubscribed = value === 'true';
                }
                else if (key === 'owner') {
                    res.owner = Strophe.getNodeFromJid(value);
                }
                else if (key === 'subscribers') {
                    res.followersSize = parseInt(value.size);
                }
                else if (key === 'location') {
                    res.location = { latitude: parseFloat(value.geoloc.lat), longitude: parseFloat(value.geoloc.lon) };
                }
                else if (key === 'updated') {
                    res.time = iso8601toDate(value).getTime();
                }
                else if (key === 'radius') {
                    res.radius = parseFloat(value);
                }
                else {
                    const numbers = ['image_items', 'total_items', 'visibility'];
                    res[camelize(key)] = numbers.indexOf(key) !== -1 ? parseInt(value) : value;
                }
            }
        }
        catch (e) {
            console.error(`Cannot process key ${key} value: ${value}`);
        }
    });
    delete res.id;
    return res;
}
exports.processMap = processMap;
function fromCamelCase(data = {}) {
    const { firstName, userID, phoneNumber, lastName, sessionID, uuid } = data, result = tslib_1.__rest(data, ["firstName", "userID", "phoneNumber", "lastName", "sessionID", "uuid"]);
    if (phoneNumber) {
        result.phone_number = phoneNumber;
        result.phoneNumber = phoneNumber;
    }
    if (userID) {
        result.auth_user = userID;
    }
    if (firstName) {
        result.first_name = firstName;
    }
    if (lastName) {
        result.last_name = lastName;
    }
    if (sessionID) {
        result.token = sessionID;
    }
    if (uuid) {
        result.user = uuid;
    }
    return result;
}
exports.fromCamelCase = fromCamelCase;
function getJid(username, host, resource) {
    let jid = `${username}@${host}`;
    if (resource) {
        jid = `${jid}/${resource}`;
    }
    return jid;
}
exports.getJid = getJid;
function getNodeJid(jid) {
    if (jid.indexOf('@') < 0) {
        return jid;
    }
    return jid.split('@')[0];
}
exports.getNodeJid = getNodeJid;
/** Function: getUniqueId
 *  Generate a unique ID for use in <iq/> elements.
 *
 *  All <iq/> stanzas are required to have unique id attributes.  This
 *  function makes creating these easy.  Each connection instance has
 *  a counter which starts from zero, and the value of this counter
 *  plus a colon followed by the suffix becomes the unique id. If no
 *  suffix is supplied, the counter is used as the unique id.
 *
 *  Suffixes are used to make debugging easier when reading the stream
 *  data, and their use is recommended.  The counter resets to 0 for
 *  every new connection for the same reason.  For connections to the
 *  same server that authenticate the same way, all the ids should be
 *  the same, which makes it easy to see changes.  This is useful for
 *  automated testing as well.
 *
 *  Parameters:
 *    (String) suffix - A optional suffix to append to the id.
 *
 *  Returns:
 *    A unique string to be used for the id attribute.
 */
function getUniqueId(suffix) {
    let uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
        let r = (Math.random() * 16) | 0, // eslint-disable-line
        v = c === 'x' ? r : (r & 0x3) | 0x8; // eslint-disable-line
        return v.toString(16);
    });
    if (typeof suffix === 'string' || typeof suffix === 'number') {
        return `${uuid}:${suffix}`;
    }
    else {
        return `${uuid}`;
    }
}
exports.getUniqueId = getUniqueId;
function generateID() {
    const time = Date.now();
    return `s${time}${pad(Math.round(Math.random() * 1000), 4)}`;
}
exports.generateID = generateID;
function parseXml(xml, arrayTags) {
    function isArray(o) {
        return Object.prototype.toString.apply(o) === '[object Array]';
    }
    function parseNode(xmlNode, res) {
        if (!xmlNode) {
            return;
        }
        if (xmlNode.nodeName === '#text') {
            /* if you want the object to have a properyty "#text" even if it is "",
                               remove that if-else and use code that is currently in else block
                               */
            if (xmlNode.nodeValue.trim() === '') {
                return;
            }
            else {
                res[xmlNode.nodeName] = xmlNode.nodeValue;
                return;
            }
        }
        const jsonNode = {};
        const existing = res[xmlNode.nodeName];
        if (existing) {
            if (!isArray(existing)) {
                res[xmlNode.nodeName] = [existing, jsonNode];
            }
            else {
                res[xmlNode.nodeName].push(jsonNode);
            }
        }
        else if (arrayTags && arrayTags.indexOf(xmlNode.nodeName) !== -1) {
            res[xmlNode.nodeName] = [jsonNode];
        }
        else {
            res[xmlNode.nodeName] = jsonNode;
        }
        if (xmlNode.attributes) {
            const len = xmlNode.attributes.length;
            for (let i = 0; i < len; i++) {
                const attribute = xmlNode.attributes[i];
                jsonNode[attribute.nodeName] = attribute.nodeValue;
            }
        }
        const length = xmlNode.childNodes.length;
        for (let i = 0; i < length; i++) {
            parseNode(xmlNode.childNodes[i], jsonNode);
        }
    }
    const result = {};
    parseNode(xml, result);
    return process(result);
}
exports.parseXml = parseXml;
// return hashcode for given string
function hashCode(s) {
    let hash = 0, i, chr, len;
    if (s.length === 0)
        return hash;
    for (i = 0, len = s.length; i < len; i++) {
        chr = s.charCodeAt(i);
        hash = (hash << 5) - hash + chr; // eslint-disable-line
        // Convert to 32bit integer.
        hash = Math.abs(hash & hash); // eslint-disable-line
    }
    return hash;
}
exports.hashCode = hashCode;
function iso8601toDate(date) {
    let timestamp = Date.parse(date), minutesOffset = 0;
    if (isNaN(timestamp)) {
        // eslint-disable-next-line
        let struct = /^(\d{4}|[+\-]\d{6})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d{3,}))?)?(?:(Z)|([+\-])(\d{2})(?::?(\d{2}))?))?/.exec(date);
        if (struct) {
            if (struct[8] !== 'Z') {
                minutesOffset = +struct[10] * 60 + +struct[11];
                if (struct[9] === '+') {
                    minutesOffset = -minutesOffset;
                }
            }
            return new Date(+struct[1], +struct[2] - 1, +struct[3], +struct[4], +struct[5] + minutesOffset, +struct[6], struct[7] ? +struct[7].substr(0, 3) : 0);
        }
        else {
            // XEP-0091 dateAsString
            timestamp = Date.parse(`${date.replace(/^(\d{4})(\d{2})(\d{2})/, '$1-$2-$3')}Z`);
        }
    }
    return new Date(timestamp);
}
exports.iso8601toDate = iso8601toDate;
//# sourceMappingURL=utils.js.map