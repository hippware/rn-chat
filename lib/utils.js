"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
function process(result) {
    if (typeof result === 'object') {
        if (Array.isArray(result)) {
            return result.map(function (el) { return process(el); });
        }
        else {
            if (result['#text'] && Object.keys(result).length === 1) {
                return result['#text'];
            }
            var res = {};
            var changed = false;
            // eslint-disable-next-line
            for (var key in result) {
                // eslint-disable-next-line
                if (result.hasOwnProperty(key)) {
                    changed = true;
                    res[key] = process(result[key]);
                }
            }
            if (!changed) {
                return undefined;
            }
            return res;
        }
    }
    else {
        return result;
    }
}
exports.default = {
    fromCamelCase: function (data) {
        if (data === void 0) { data = {}; }
        var firstName = data.firstName, userID = data.userID, phoneNumber = data.phoneNumber, lastName = data.lastName, sessionID = data.sessionID, uuid = data.uuid, result = tslib_1.__rest(data, ["firstName", "userID", "phoneNumber", "lastName", "sessionID", "uuid"]);
        if (phoneNumber) {
            result.phone_number = phoneNumber;
            result.phoneNumber = phoneNumber;
        }
        if (userID) {
            result.auth_user = userID;
        }
        if (firstName) {
            result.first_name = firstName;
        }
        if (lastName) {
            result.last_name = lastName;
        }
        if (sessionID) {
            result.token = sessionID;
        }
        if (uuid) {
            result.user = uuid;
        }
        return result;
    },
    getJid: function (username, host, resource) {
        var jid = username + "@" + host;
        if (resource) {
            jid = jid + "/" + resource;
        }
        return jid;
    },
    getNodeJid: function (jid) {
        if (jid.indexOf('@') < 0) {
            return null;
        }
        return jid.split('@')[0];
    },
    /** Function: getUniqueId
     *  Generate a unique ID for use in <iq/> elements.
     *
     *  All <iq/> stanzas are required to have unique id attributes.  This
     *  function makes creating these easy.  Each connection instance has
     *  a counter which starts from zero, and the value of this counter
     *  plus a colon followed by the suffix becomes the unique id. If no
     *  suffix is supplied, the counter is used as the unique id.
     *
     *  Suffixes are used to make debugging easier when reading the stream
     *  data, and their use is recommended.  The counter resets to 0 for
     *  every new connection for the same reason.  For connections to the
     *  same server that authenticate the same way, all the ids should be
     *  the same, which makes it easy to see changes.  This is useful for
     *  automated testing as well.
     *
     *  Parameters:
     *    (String) suffix - A optional suffix to append to the id.
     *
     *  Returns:
     *    A unique string to be used for the id attribute.
     */
    getUniqueId: function (suffix) {
        var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = (Math.random() * 16) | 0, // eslint-disable-line
            v = c === 'x' ? r : (r & 0x3) | 0x8; // eslint-disable-line
            return v.toString(16);
        });
        if (typeof suffix === 'string' || typeof suffix === 'number') {
            return uuid + ":" + suffix;
        }
        else {
            return "" + uuid;
        }
    },
    parseXml: function (xml, arrayTags) {
        function isArray(o) {
            return Object.prototype.toString.apply(o) === '[object Array]';
        }
        function parseNode(xmlNode, res) {
            if (!xmlNode) {
                return;
            }
            if (xmlNode.nodeName === '#text') {
                /* if you want the object to have a properyty "#text" even if it is "",
                                 remove that if-else and use code that is currently in else block
                                 */
                if (xmlNode.nodeValue.trim() === '') {
                    return;
                }
                else {
                    res[xmlNode.nodeName] = xmlNode.nodeValue;
                    return;
                }
            }
            var jsonNode = {};
            var existing = res[xmlNode.nodeName];
            if (existing) {
                if (!isArray(existing)) {
                    res[xmlNode.nodeName] = [existing, jsonNode];
                }
                else {
                    res[xmlNode.nodeName].push(jsonNode);
                }
            }
            else if (arrayTags && arrayTags.indexOf(xmlNode.nodeName) !== -1) {
                res[xmlNode.nodeName] = [jsonNode];
            }
            else {
                res[xmlNode.nodeName] = jsonNode;
            }
            if (xmlNode.attributes) {
                var len = xmlNode.attributes.length;
                for (var i = 0; i < len; i++) {
                    var attribute = xmlNode.attributes[i];
                    jsonNode[attribute.nodeName] = attribute.nodeValue;
                }
            }
            var length = xmlNode.childNodes.length;
            for (var i = 0; i < length; i++) {
                parseNode(xmlNode.childNodes[i], jsonNode);
            }
        }
        var result = {};
        parseNode(xml, result);
        return process(result);
    },
    // return hashcode for given string
    hashCode: function (s) {
        var hash = 0, i, chr, len;
        if (s.length === 0)
            return hash;
        for (i = 0, len = s.length; i < len; i++) {
            chr = s.charCodeAt(i);
            hash = (hash << 5) - hash + chr; // eslint-disable-line
            // Convert to 32bit integer.
            hash = Math.abs(hash & hash); // eslint-disable-line
        }
        return hash;
    },
    iso8601toDate: function (date) {
        var timestamp = Date.parse(date), minutesOffset = 0;
        if (isNaN(timestamp)) {
            // eslint-disable-next-line
            var struct = /^(\d{4}|[+\-]\d{6})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.(\d{3,}))?)?(?:(Z)|([+\-])(\d{2})(?::?(\d{2}))?))?/.exec(date);
            if (struct) {
                if (struct[8] !== 'Z') {
                    minutesOffset = +struct[10] * 60 + +struct[11];
                    if (struct[9] === '+') {
                        minutesOffset = -minutesOffset;
                    }
                }
                return new Date(+struct[1], +struct[2] - 1, +struct[3], +struct[4], +struct[5] + minutesOffset, +struct[6], struct[7] ? +struct[7].substr(0, 3) : 0);
            }
            else {
                // XEP-0091 dateAsString
                timestamp = Date.parse(date.replace(/^(\d{4})(\d{2})(\d{2})/, '$1-$2-$3') + "Z");
            }
        }
        return new Date(timestamp);
    }
};
//# sourceMappingURL=utils.js.map