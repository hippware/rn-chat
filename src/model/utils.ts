import {when} from 'mobx'

export async function waitFor(condition: () => boolean) {
  return new Promise((resolve, reject) => {
    when(
      () => {
        let res = false
        try {
          res = condition()
        } catch (e) {
          reject(e)
        }
        return res
      },
      () => {
        resolve()
      }
    )
  })
}
