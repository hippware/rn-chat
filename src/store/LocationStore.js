// @flow

import {types, getEnv, flow, getParent} from 'mobx-state-tree';
import {when, autorun, reaction} from 'mobx';
import Permissions from 'react-native-permissions';
import {settings} from '../globals';

export const Location = types.model('Location', {
  longitude: types.number,
  latitude: types.number,
  accuracy: types.number,
});

const METRIC = 'METRIC';
const IMPERIAL = 'IMPERIAL';

const LocationStore = types
  .model('LocationStore', {})
  .volatile(self => ({
    // TODO: should we persist location?
    location: types.maybe(Location),
    enabled: true,
    alwaysOn: true,
    system: types.optional(types.enumeration('Metric system', [METRIC, IMPERIAL]), METRIC),
    loading: false,
  }))
  .views(self => ({
    distance: (lat1, lon1, lat2, lon2) => {
      var R = 6371000; // Radius of the earth in m
      var dLat = (lat2 - lat1) * Math.PI / 180; // deg2rad below
      var dLon = (lon2 - lon1) * Math.PI / 180;
      var a = 0.5 - Math.cos(dLat) / 2 + Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * (1 - Math.cos(dLon)) / 2;

      const res = R * 2 * Math.asin(Math.sqrt(a));
      const result = self.system === METRIC ? res : res * 3.2808399;
      return result;
    },

    distanceToString: (distance) => {
      const limit = self.system === METRIC ? 1000 : 5280;
      // if (distance>limit){
      return self.system === METRIC ? `${Math.round(distance / 100) / 10} km` : `${Math.round(distance * 0.00189393939) / 10} mi`;
      // } else {
      //   return this.system === METRIC ? `${Math.trunc(distance)} m` : `${Math.trunc(distance/0.3048)} ft`;
      // }
    },

    distanceFromBot: (botLoc: {latitude: number, longitude: number}) => {
      const {location, distanceToString, distance} = self;
      if (location && botLoc) {
        return distanceToString(distance(location.latitude, location.longitude, botLoc.latitude, botLoc.longitude));
      }
      return null;
    },
  }))
  .actions((self) => {
    const {logger, geolocation, nativeEnv} = getEnv(self);
    let wocky, watch, handler;

    function afterAttach() {
      ({wocky} = getParent(self));
      self.initialize();
      self.start();
    }
    function start() {
      Permissions.check('location', {type: 'always'}).then((response) => self.setAlwaysOn(response === 'authorized'));
      handler = reaction(
        () => wocky.connected,
        () => {
          self.startBackground();
        },
      );
      self.getCurrentPosition();
    }

    function finish() {
      if (!self.alwaysOn) {
        self.stopBackground();
      }
      handler();
    }

    function startBackground() {
      const model = getParent(self).wocky;
      const {backgroundGeolocation, backgroundFetch} = getEnv(self);
      if (typeof navigator !== 'undefined') {
        logger.log('BACKGROUND LOCATION START');

        backgroundFetch.configure(
          {
            stopOnTerminate: false,
          },
          () => {
            logger.log('background-fetch: Received background-fetch event');

            // todo: explicitly call for an update on position?

            // To signal completion of your task to iOS, you must call #finish!
            // If you fail to do this, iOS can kill your app.
            backgroundFetch.finish();
          },
          (error) => {
            logger.log('[js] RNBackgroundFetch failed to start');
          },
        );

        // // This handler fires whenever bgGeo receives a location update.
        backgroundGeolocation.on('location', (position) => {
          logger.log('- [js]location: ', JSON.stringify(position));
          self.setPosition(position);
          // this.location = position.coords;
          // we don't need it because we have HTTP location share
          // this.share(this.location);
        });

        // This handler fires when movement states changes (stationary->moving; moving->stationary)
        backgroundGeolocation.on('http', (response) => {
          logger.log('- [js]http: ', response.responseText);
          //        logger.log('- [js]http: ', JSON.parse(response.responseText));
        });

        // This handler fires whenever bgGeo receives an error
        backgroundGeolocation.on('error', (error) => {
          self.positionError(error);
          // var type = error.type;
          // var code = error.code;
          // alert(type + " Error: " + code);
        });

        // This handler fires when movement states changes (stationary->moving; moving->stationary)
        backgroundGeolocation.on('motionchange', (location) => {
          logger.log('- [js]motionchanged: ', JSON.stringify(location));
        });

        // This event fires when a chnage in motion activity is detected
        backgroundGeolocation.on('activitychange', (activityName) => {
          logger.log('- Current motion activity: ', activityName); // eg: 'on_foot', 'still', 'in_vehicle'
        });

        // This event fires when the user toggles location-services
        backgroundGeolocation.on('providerchange', (provider) => {
          logger.log('- Location provider changed: ', provider);

          self.setAlwaysOn(provider.status === backgroundGeolocation.AUTHORIZATION_STATUS_ALWAYS);
        });
        const url = `https://${settings.getDomain()}/api/v1/users/${model.username}/locations`;
        logger.log(`LOCATION UPDATE URL: ${url} ${model.username} ${model.password}`);
        backgroundGeolocation.configure(
          {
            // Geolocation Config
            desiredAccuracy: 0,
            useSignificantChangesOnly: false,
            stationaryRadius: 20,
            distanceFilter: 30,
            // Activity Recognition
            stopTimeout: 1,
            // Application config
            debug: false, // <-- enable this hear sounds for background-geolocation life-cycle.
            // logLevel: backgroundGeolocation.LOG_LEVEL_VERBOSE,
            stopOnTerminate: false, // <-- Allow the background-service to continue tracking when user closes the app.
            startOnBoot: true, // <-- Auto start tracking when device is powered-up.
            // HTTP / SQLite config
            url,
            // batchSync: false, // <-- [Default: false] Set true to sync locations to server in a single HTTP request.
            autoSync: true, // <-- [Default: true] Set true to sync each location to server as it arrives.
            // maxDaysToPersist: 1, // <-- Maximum days to persist a location in plugin's SQLite database when HTTP fails
            headers: {
              // <-- Optional HTTP headers
              'X-Auth-User': model.username,
              'X-Auth-Token': model.password,
            },
            params: {
              // <-- Optional HTTP params
              resource: 'testing',
            },
          },
          (state) => {
            logger.log('- BackgroundGeolocation is configured and ready: ', state);

            if (!state.enabled && self.alwaysOn) {
              backgroundGeolocation.start(() => {
                logger.log('- Start success');
              });
            } else if (state.enabled && !self.alwaysOn) {
              backgroundGeolocation.stop();
            }
          },
        );
      }
    }

    function stopBackground() {
      const {backgroundGeolocation} = getEnv(self);
      if (typeof backgroundGeolocation !== 'undefined') {
        backgroundGeolocation.stop();
      }
    }

    function initialize() {
      const system = nativeEnv.get('NSLocaleUsesMetricSystem') ? 'METRIC' : 'IMPERIAL';
      setMetricSystem(system);
    }

    function setMetricSystem(type) {
      self.system = type;
    }

    const getCurrentPosition = flow(function* getCurrentPosition() {
      if (self.loading) return self.location;
      self.loading = true;
      yield new Promise((resolve, reject) => {
        navigator.geolocation.getCurrentPosition(
          (position) => {
            self.setPosition(position);
            resolve();
          },
          (error) => {
            self.positionError(error);
            reject();
          },
          {timeout: 20000, maximumAge: 1000, enableHighAccuracy: false},
        );
      });
    });

    function setPosition(position) {
      self.enabled = true;
      self.location = position.coords;
      self.loading = false;
      // TODO: share location via wocky-client
      // this.share(this.location);
    }

    function positionError(error) {
      if (error.code === 1) {
        // user denied location permissions
        self.enabled = false;
      }
      self.loading = false;
      // @TODO: how do we handle timeout or other error?
      logger.log('LOCATION ERROR:', error, error.message, {level: logger.levels.ERROR});
    }

    function watchPosition() {
      if (!watch) {
        watch = geolocation.watchPosition(self.setPosition, self.positionError, {
          timeout: 20000,
          maximumAge: 60000,
          enableHighAccuracy: false,
        });
      }
    }

    function setAlwaysOn(value) {
      self.alwaysOn = value;
    }

    return {start, finish, afterAttach, setAlwaysOn, watchPosition, setPosition, stopBackground, startBackground, positionError, getCurrentPosition, initialize};
  });

export default LocationStore;
