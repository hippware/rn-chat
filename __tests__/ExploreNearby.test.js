// @flow
/* global test */

import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'mobx-react/native';
import LocationStore from '../src/store/LocationStore';
import injects from './utils/inject-mocks';
import './utils/mockMap';
import ExploreNearby from '../src/components/map/ExploreNearBy';

describe('Explore Nearby', () => {
  test('renders with no data', () => {
    const wocky = {};

    const nativeEnv = {
      get: () => null,
    };

    const geolocation = {
      watchPosition: () => {},
      getCurrentPosition: () => {},
    };

    const locationStore = LocationStore.create({}, {geolocation, nativeEnv, logger: {log: () => {}}});
    locationStore.setPosition({
      coords: {longitude: 2, latitude: 2, accuracy: 2},
    });

    const toRender = (
      <Provider {...injects} wocky={wocky} locationStore={locationStore}>
        <ExploreNearby />
      </Provider>
    );
    const tree = renderer.create(toRender).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
