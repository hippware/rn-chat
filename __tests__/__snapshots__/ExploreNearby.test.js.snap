// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore Nearby renders with no data 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <MapView
      animateClusters={true}
      autoZoom={true}
      clusterInitialDimension={30}
      clusterInitialFontSize={12}
      clusterPressMaxChildren={100}
      clusteringEnabled={true}
      containerStyle={Object {}}
      customMapStyle={
        Array [
          Object {
            "elementType": "labels.text",
            "featureType": "administrative",
            "stylers": Array [
              Object {
                "color": "#fe5c6c",
              },
            ],
          },
          Object {
            "elementType": "labels.text.stroke",
            "featureType": "administrative",
            "stylers": Array [
              Object {
                "color": "#ffffff",
              },
            ],
          },
          Object {
            "elementType": "labels.text",
            "featureType": "administrative.land_parcel",
            "stylers": Array [
              Object {
                "color": "#d6949b",
              },
            ],
          },
          Object {
            "elementType": "labels.text.fill",
            "featureType": "administrative.neighborhood",
            "stylers": Array [
              Object {
                "color": "#d6949b",
              },
            ],
          },
          Object {
            "elementType": "geometry.fill",
            "featureType": "landscape.man_made",
            "stylers": Array [
              Object {
                "color": "#fff8f0",
              },
            ],
          },
          Object {
            "elementType": "geometry.fill",
            "featureType": "landscape.natural",
            "stylers": Array [
              Object {
                "color": "#fff8f0",
              },
            ],
          },
          Object {
            "elementType": "geometry.stroke",
            "featureType": "road",
            "stylers": Array [
              Object {
                "visibility": "off",
              },
            ],
          },
          Object {
            "elementType": "labels.text",
            "featureType": "road",
            "stylers": Array [
              Object {
                "color": "#70b0e1",
              },
            ],
          },
          Object {
            "elementType": "labels.text.stroke",
            "featureType": "road",
            "stylers": Array [
              Object {
                "color": "#ffffff",
              },
            ],
          },
          Object {
            "elementType": "labels.icon",
            "featureType": "road.highway",
            "stylers": Array [
              Object {
                "visibility": "off",
              },
            ],
          },
          Object {
            "featureType": "transit",
            "stylers": Array [
              Object {
                "visibility": "off",
              },
            ],
          },
          Object {
            "elementType": "geometry.fill",
            "featureType": "water",
            "stylers": Array [
              Object {
                "color": "#99ddf0",
              },
            ],
          },
        ]
      }
      data={Array []}
      edgePadding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      extent={512}
      followUser={true}
      fullMap={true}
      height={1334}
      initialRegion={
        Object {
          "latitude": 2,
          "latitudeDelta": 0.04,
          "longitude": 2,
          "longitudeDelta": 0.04,
        }
      }
      locationStore={
        Object {
          "backgroundOptions": Object {
            "activityRecognitionInterval": null,
            "activityType": null,
            "debug": null,
            "desiredAccuracy": null,
            "distanceFilter": null,
            "elasticityMultiplier": null,
            "stationaryRadius": null,
            "stopTimeout": null,
          },
          "fetchResult": 0,
          "location": Object {
            "accuracy": 2,
            "latitude": 2,
            "longitude": 2,
          },
        }
      }
      maxZoom={20}
      minZoom={1}
      onMarkerPress={[Function]}
      onPress={[Function]}
      onRegionChangeComplete={[Function]}
      preserveClusterPressBehavior={true}
      provider="google"
      renderCluster={[Function]}
      renderMarker={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      textStyle={Object {}}
      width={750}
      wocky={Object {}}
    >
      <Marker
        coordinate={
          Object {
            "latitude": 2,
            "longitude": 2,
          }
        }
        style={
          Object {
            "zIndex": 1000,
          }
        }
        tracksViewChanges={true}
      >
        <View
          style={
            Object {
              "transform": Array [],
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../images/location-indicator.png",
              }
            }
          />
        </View>
      </Marker>
    </MapView>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 40,
          "height": 100,
          "left": 30,
          "opacity": 1,
          "position": "absolute",
          "width": 100,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../images/currentLocationGrey.png",
          }
        }
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgb(254,92,108)",
          "borderRadius": 27,
          "bottom": 40,
          "height": 54,
          "opacity": 1,
          "position": "absolute",
          "right": 40,
          "shadowOffset": Object {
            "height": 0,
            "width": 2,
          },
          "shadowOpacity": 0.18,
          "shadowRadius": 4,
          "width": 54,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../images/iconCreateBot.png",
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#007bff",
          "borderBottomWidth": 1,
          "borderColor": "#006acd",
          "bottom": null,
          "left": 0,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": -1334,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
                "marginLeft": 10,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
